### PR TITLE
OSRA412 - Add sponsorships#create controller and sponsorships#new view

### PR DESCRIPTION
**Pull request by gmunumel**
OSRA412 - Add sponsorships#create controller and sponsorships#new view
https://osraav.atlassian.net/browse/OSRA-412

### CODE REVIEW

**Comment by gmunumel** [comment]
WIP - Still missing some functionality

osra
osra3

**Comment by gmunumel** [comment]
**Sponsorship create**
**Scope elegible for sponsorship**
![screenshot from 2015-05-19 23 57 50]
(https://cloud.githubusercontent.com/assets/276168/7715797/3b1c1ff8-fe83-11e4-9e61-42d1e90929c7.png)

**Scope all**
screenshot from 2015-05-19 23 58 22

File: app/controllers/hq/orphans_controller.rb [hunk 1]
```diff
@@ -23,6 +24,17 @@ def update
 
 private
 
+  def load_scope
+    @orphans_all = Orphan.all().size
```
Comment (razorcd) [inline app/controllers/hq/orphans_controller.rb]: `Orphan.count` is better even if both invoke the same query.

File: app/controllers/hq/sponsorships_controller.rb [hunk 2]
```diff
@@ -1,5 +1,19 @@
 class Hq::SponsorshipsController < HqController
 
+  def create
+    params.permit :orphan_id
```
Comment (razorcd) [inline app/controllers/hq/sponsorships_controller.rb]: is L4 doing anything here?

File: app/controllers/hq/sponsorships_controller.rb [hunk 3]
```diff
@@ -1,5 +1,19 @@
...
+    sponsor = Sponsor.find(params[:sponsor_id])
+    orphan = Orphan.find(params[:orphan_id])
+    sponsorship = sponsor.sponsorships.build(orphan_id: params[:orphan_id],
+                                             start_date: params[:sponsorship_start_date])
+
+
```
Comment (razorcd) [inline app/controllers/hq/sponsorships_controller.rb]: remove empty line 10

File: app/views/hq/orphans/index.html.erb [hunk 4]
```diff
@@ -90,6 +133,21 @@
...
+                    <input name="_method" type="hidden" value="post">
+                    <input name="utf8" type="hidden" value="&#x2713;" />
+                    <input name="authenticity_token" type="hidden" value="<%= form_authenticity_token %>" />
+                    <div class="form-group">
+                      <label for="title">Beginning</label>
+                      <input name="sponsorship_start_date" type="text" value="<%= Date.current.to_s %>" class="form-control date-picker">
+                      <input name="orphan_id" type="hidden" value="<%= orphan.id %>" />
+                      <input type="submit" value="Sponsor this orphan" class="btn btn-primary" />
+                    </div>
+                  </form>
```
Comment (razorcd) [inline app/views/hq/orphans/index.html.erb]: why not use `form_tag` instead of plain html?
I think `action` should be `hq_sponsor_sponsorships(@sponsor)`. check the `rake routes` to be sure

**Comment by razorcd** [comment]
looks like the controller and view tests are missing

File: app/controllers/hq/orphans_controller.rb (hunk 1 above)
Comment (gmunumel) [inline app/controllers/hq/orphans_controller.rb]: I've read [this](http://web.archive.org/web/20100210204319/http://blog.hasmanythrough.com/2008/2/27/count-length-size) article, and stay that if I use `size` I don't mind to care about the choose, because it changes between the state in that moment. I'd like to take the more efficient option tho.

File: app/controllers/hq/sponsorships_controller.rb (hunk 2 above)
Comment (gmunumel) [inline app/controllers/hq/sponsorships_controller.rb]: I got an error if I don't define a `permit` params :(

File: app/controllers/hq/sponsorships_controller.rb (hunk 3 above)
Comment (gmunumel) [inline app/controllers/hq/sponsorships_controller.rb]: Sure thing :+1:

File: app/views/hq/orphans/index.html.erb (hunk 4 above)
Comment (gmunumel) [inline app/views/hq/orphans/index.html.erb]: I've tried, but got stuck with the action part, as you guessed. I'll give another shot.

File: app/controllers/hq/sponsorships_controller.rb [hunk 5]
```diff
@@ -20,6 +33,23 @@ def destroy
...
+        redirect_to hq_sponsor_path(sponsor)
+      else
+        flash[:success] = "Sponsorship link was successfully created for #{orphan.full_name}"
+        redirect_to hq_new_sponsorship_path(sponsor, scope: 'eligible_for_sponsorship')
+      end
+    end
+  end
+
+  def redirect_back_to_new_sponsorship_for(sponsor)
+    flash[:warning] = @sponsorship_creator.error_msg
```
Comment (razorcd) [inline app/controllers/hq/sponsorships_controller.rb]: this should be `flash[:error]`
Also add some steps to the integration tests to check for sad paths too !

Comment (razorcd) [inline app/controllers/hq/sponsorships_controller.rb]: Sad paths being:
- for sponsorship start_date validation
- for sponsors that are inactive or on hold can't create sponsorships

File: app/controllers/hq/orphans_controller.rb [hunk 6]
```diff
@@ -23,6 +24,18 @@ def update
 
 private
 
+  def load_scope
+    @orphans_all = Orphan.count
+    @orphans_sort_by_eligibility = Orphan.sort_by_eligibility.count
```
Comment (razorcd) [inline app/controllers/hq/orphans_controller.rb]: Names of vars are confusing.
Maybe `orphans_count`  and `orphans_sort_by_eligibility_count`?

File: config/routes.rb [hunk 7]
```diff
@@ -15,12 +15,15 @@
...
+      resources :sponsorships, only: [:create, :destroy] , shallow: true do
         put "inactivate", on: :member
       end
     end
     resources :orphans, except: [:new, :create, :destroy]
     resources :admin_users, except: [:show]
+
+    #build a new sponsorship on the Orphan class instead of the Sponsorship class
+    get '/sponsors/:sponsor_id/sponsorships/new', to: 'orphans#index', as: :new_sponsorship
   end
```
Comment (razorcd) [inline config/routes.rb]: Also add tests for all these new roots in `spec/routing`

File: spec/controllers/hq/orphans_controller_spec.rb [hunk 8]
```diff
@@ -14,7 +14,7 @@
 
   specify '#index' do
     expect(Orphan).to receive(:paginate).with(page: "2")
-      .and_return(orphans.paginate(per_page: 2, page: 1))
+                          .and_return(orphans.paginate(per_page: 2, page: 1))
```
Comment (razorcd) [inline spec/controllers/hq/orphans_controller_spec.rb]: don't make unnecessary changes like these, it might cause merge conflicts for others.

File: spec/controllers/hq/sponsorships_controller_spec.rb [hunk 9]
```diff
@@ -2,9 +2,81 @@
...
+      context 'when request was fulfilled' do
+        before(:each) do
+          expect(sponsor).to receive(:request_fulfilled).and_return true
+
+          post :create, { id: 1, orphan_id: 1, sponsor_id: 1, start_date: '1-1-2013' }
+        end
+
+        it 'sets flash[:success] message' do
+          expect(flash[:success]).not_to be_nil
+          expect(flash[:warning]).to be_nil
```
Comment (razorcd) [inline spec/controllers/hq/sponsorships_controller_spec.rb]: `:error`

File: spec/controllers/hq/sponsorships_controller_spec.rb [hunk 10]
```diff
@@ -2,9 +2,81 @@
...
+        before(:each) do
+          expect(sponsor).to receive(:request_fulfilled).and_return false
+          expect(orphan).to receive(:full_name).and_return 'first orphan'
+
+          post :create, { id: 1, orphan_id: 1, sponsor_id: 1, start_date: '1-1-2013' }
+        end
+
+        it 'sets flash[:warning] message' do
+          expect(flash[:success]).not_to be_nil
+          expect(flash[:warning]).to be_nil
```
Comment (razorcd) [inline spec/controllers/hq/sponsorships_controller_spec.rb]: `:error` on L54 and L52

File: spec/controllers/hq/sponsorships_controller_spec.rb [hunk 11]
```diff
@@ -2,9 +2,81 @@
...
+      before(:each) do
+        expect(sponsorship_creator).to receive(:call).and_return false
+        expect(sponsorship_creator).to receive(:error_msg).and_return 'No go'
+
+        post :create, { id: 1, orphan_id: 1, sponsor_id: 1, start_date: '1-1-2013' }
+      end
+
+      it 'sets flash[:warning] message' do
+        expect(flash[:success]).to be_nil
+        expect(flash[:warning]).to eq 'No go'
```
Comment (razorcd) [inline spec/controllers/hq/sponsorships_controller_spec.rb]: `:error` L73 and L71

File: spec/features/hq/create_sponsorship_spec.rb [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+  end
+
+  def then_i_should_be_on(page, sponsor_name)
+    sponsor = Sponsor.find_by name: sponsor_name
+    case page.to_sym
+      when :hq_sponsor_path then expect(current_path).to eq hq_sponsor_path sponsor.id
+      when :hq_new_sponsorship_path then expect(current_path).to eq hq_new_sponsorship_path sponsor.id
+      else raise('path to specified is not listed in #path_to')
+    end
+  end
```
Comment (razorcd) [inline spec/features/hq/create_sponsorship_spec.rb]: we already have this in `spec/support/feature_helpers/navigation.rb` . Edit there.

File: spec/controllers/hq/orphans_controller_spec.rb (hunk 8 above)
Comment (gmunumel) [inline spec/controllers/hq/orphans_controller_spec.rb]: Got it!

File: spec/features/hq/create_sponsorship_spec.rb (hunk 12 above)
Comment (gmunumel) [inline spec/features/hq/create_sponsorship_spec.rb]: @razorcd Do we have a special functionality to activate `Sponsor`s?  I'm trying to create a sponsor with status `active` but is always created `inactive`.

Comment (gmunumel) [inline spec/features/hq/create_sponsorship_spec.rb]: @razorcd but in this case i'm using an extra argument, `sponsor_name`. Do I need to added to `navigation.rb` anyway?

File: app/controllers/hq/sponsorships_controller.rb (hunk 2 above)
Comment (razorcd) [inline app/controllers/hq/sponsorships_controller.rb]: This line should be removed. We are not doing mass assignment so there is no need for this.

File: app/views/hq/orphans/_form.html.erb [hunk 13]
```diff
@@ -116,7 +116,7 @@
       </div>
       <div class="form-group">
         <%= f.label_for_field :guardian_id_num %>
-        <%= f.text_field :guardian_id_num, class: "form-control" %>
+        <%= f.number_field :guardian_id_num, class: "form-control" %>
```
Comment (razorcd) [inline app/views/hq/orphans/_form.html.erb]: The `number_field` is more useful when it is used for fields that represent some kind of quantity.
Also changes to the behavior of the application should be discussed with the client and dev team first.
Also this is not related to the current task.
I would prefer to keep this as it was.

File: app/views/hq/sponsors/show.html.haml [hunk 14]
```diff
@@ -2,8 +2,12 @@
   %div.instance
     = link_to 'Edit Sponsor', edit_hq_sponsor_path(@sponsor.id),
       class: 'btn btn-default', role: 'button'
-    = link_to 'Link to Orphan', '#',
+    = link_to 'New Sponsor', new_hq_sponsor_path,
       class: 'btn btn-default', role: 'button'
+    - if @sponsor.status.name != 'Inactive' and @sponsor.status.name != 'On Hold'
```
Comment (razorcd) [inline app/views/hq/sponsors/show.html.haml]: I think `if @sponsor.eligible_for_sponsorship?` would be better here.
Adding an `else` and show the button `disabled` would be useful

File: spec/features/hq/create_sponsorship_spec.rb [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+    and_i_should_be_on "hq_new_sponsorship_path", "First Sponsor"
+    and_i_should_see "First Sponsor"
+    and_i_should_see "First Orphan"
+    and_i_should_see "Second Orphan"
+    fill_orphan_with "sponsorship_start_date", "First Orphan", "2014-01-01"
+    when_i_click "Sponsor this orphan", "First Orphan"
+    and_i_should_be_on "hq_new_sponsorship_path", "First Sponsor"
+    and_i_should_see "Return to Sponsor Page"
+    and_i_should_see "No (1/2)"
+    and_i_should_see "Sponsorship link was successfully created for First Orphan"
```
Comment (razorcd) [inline spec/features/hq/create_sponsorship_spec.rb]: add after this line a step that fails on sponsorship validation:
```
fill_orphan_with "sponsorship_start_date", "Second Orphan", "2100-02-01"
when_i_click "Sponsor this orphan", "Second Orphan"
and_i_should_see "Validation failed"
```

File: spec/features/hq/create_sponsorship_spec.rb [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+    fill_orphan_with "sponsorship_start_date", "Second Orphan", "2014-02-01"
+    when_i_click "Sponsor this orphan", "Second Orphan"
+    and_i_should_be_on "hq_sponsor_path", "First Sponsor"
+    and_i_should_see_within "Currently Sponsored Orphans", "First Orphan"
```
Comment (razorcd) [inline spec/features/hq/create_sponsorship_spec.rb]: add: `and_i_should_see "Sponsorship link was successfully created"`
to also test the presence of the flash message

File: spec/support/feature_helpers/navigation.rb [hunk 17]
```diff
@@ -1,10 +1,13 @@
...
+    def and_i_should_be_on page, sponsor_name = ''
+      sponsor = Sponsor.find_by(name: sponsor_name) if not sponsor_name.empty?
       page_path = case page.to_sym
                     when :root_page then hq_root_path
                     when :sign_in_page then new_hq_admin_user_session_path
                     when :new_sponsor_page then new_hq_sponsor_path
+                    when :hq_sponsor_path then hq_sponsor_path sponsor.id
+                    when :hq_new_sponsorship_path then hq_new_sponsorship_path sponsor.id
                     else raise('path to specified is not listed in #path_to')
                   end
```
Comment (razorcd) [inline spec/support/feature_helpers/navigation.rb]: This is not very reusable. If we need it to use it for `orpahns_name` it will be a problem.
Let's refactor this to respect the SOLID design principles.


One option would be to rename this `def and_i_should_be_on path` to  `def get_path path, id`

Then to handle the conversion from name to id:
``` ruby
#use like:  and_i_should_be_on "hq_sponsor_path", sponsor_name: "Sponsor 1"
def and_i_should_be_on path, field_hash
     case  field_hash.keys.first.to_sym
          when :sponsor_name do
                 sponsor = Sponsor.find_by(name: field_hash[:sponsor_name])
                 get_path path, sponsor.id
         end
          when :sponsor_id do
                 get_path path, field_hash[:sponsor_id]
         end
         else raise('some error message')
   end

end
```
(this wasn't tested, i got carried away, refactor where needed)

File: app/views/hq/orphans/_form.html.erb (hunk 13 above)
Comment (gmunumel) [inline app/views/hq/orphans/_form.html.erb]: Yes, I agree. But I got an error related to this field in the form view validation. It complains that the view doesn't have the field: https://github.com/AgileVentures/osra/blob/develop/spec/views/hq/orphans/_form_spec.rb#L47

**Comment by gmunumel** [comment]
At the end the `permit_params` wasn't neccesary. Still thinking about what could be the cause of the error before.

**Comment by razorcd** [comment]
:+1:
